src/shogun/mathematics/graph/GraphEnv.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben, Viktor Gal
+ */
+
+#include "GraphEnv.h"
+#include "GraphExecutor.h"
+
+#include <shogun/io/SGIO.h>
+
+#include <regex>
+
+using namespace shogun::graph;
+
+
+GRAPH_BACKEND GraphEnv::get_backend() const
+{
+	return m_backend;
+}
+
+void GraphEnv::set_backend(const GRAPH_BACKEND backend)
+{
+	if (available_backends().count(backend))
+	{
+		// mutex is locked here to avoid having graphs with mixed backend ops
+		std::lock_guard<std::mutex> lock(m_env_mutex);
```

Comment:
@vigsterkr this is the other place where the mutex can be locked

Reply:
how about using shared/unique locks i.e. shared_mutex? https://en.cppreference.com/w/cpp/thread/shared_mutex

Reply:
ah good point, I had in my mind this was C++20!